File: 2685-first-completely-painted-row-or-column/2685-first-completely-painted-row-or-column.py
class Solution:
    def firstCompleteIndex(self, arr: List[int], mat: List[List[int]]) -> int:
        mp = {}
        rows = len(mat)
        cols = len(mat[0])
        col_count = [rows] * cols
        row_count = [cols] * rows

        for i in range(rows):
            for j in range(cols):
                mp[mat[i][j]] = (i, j)
        
        for i in range(len(arr)):
            row, col = mp[arr[i]]
            row_count[row] -= 1
            col_count[col] -= 1
            if row_count[row] == 0 or col_count[col] == 0:
                return i

        return -1

File: 2685-first-completely-painted-row-or-column/2685-first-completely-painted-row-or-column.cpp
class Solution {
public:
    int firstCompleteIndex(vector<int>& arr, vector<vector<int>>& mat) {
        int m = mat.size();
        int n = mat[0].size();
        unordered_map<int, pair<int, int>> mp;
        vector<int> row_count(m, n), col_count(n, m);

        for( int i = 0; i < m; i++){
            for ( int j = 0; j < n; j++){
                int key = mat[i][j];
                mp[key] = {i, j};
            }
        }

        for ( int i = 0; i < arr.size(); i++){
            auto [row, col] = mp[arr[i]];
            if ( --row_count[row] == 0 || --col_count[col] == 0 ){
                return i;
            }
        }

        return -1;
    }
};